Support geo label position as runtime field

craigtaverner opened the pull request:
There is a need to provide sensibly calculated label positions for polygons and lines in Kibana maps (as requested in https://github.com/elastic/elasticsearch/issues/86044). A very convenient way of satisfy this need is through a runtime field that the rest API can make use of when labels are requested. This has the advantage of providing painless access to the label position as well. This PR provides the runtime field.

craigtaverner committed:
Support geo label position as runtime field

craigtaverner committed:
Update docs/changelog/86154.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Select first point or line segment for label

craigtaverner committed:
Fixed issues with precommit

craigtaverner commented:
@elasticmachine run elasticsearch-ci/part-2

craigtaverner committed:
Code review feedback, don't use centroid

We simplified this to always use the middle of the first tree geometry.
The use of the centroid can be done higher up in the stack.
For painless, this means in the user written script.
For the REST API we will add that logic on the server side.

craigtaverner committed:
Fixed yaml test for label position

Now that we no longer use the centroid for polygon labels, but the
middle of the first triangle.

craigtaverner committed:
Fixed yaml test for label position

This time without rounding error

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

craigtaverner committed:
Use centroid for label position on polygons

And more tests, including testing intersects for all label positions.

iverase commented on server/src/main/java/org/elasticsearch/script/field/GeoPointDocValuesField.java:
```diff
@@ -121,6 +121,11 @@ public GeoBoundingBox getInternalBoundingBox() {
 
+    @Override
+    public GeoPoint getInternalLabelPosition() {
+        return centroid;
```
I think we should return here one of the points in the array instead of the centroid.

craigtaverner replied:
Good point. I notice that the centroid is calculated during setMultiValue using a loop over all points. Since we are looping there anyway, perhaps we could pick a point somewhat central as the label position already there?

craigtaverner replied:
I think I came up with a reasonable algorithm for picking which point to use, and the index of the point in the values array is selected during the setSingleValue and setMultiValue methods, so hopefully has negligable impact. I'll push something soon.

craigtaverner committed:
Fixed multi-point to not use centroid

The centroid is unlikely to be one of the points, so we choose a point
closest to the middle of the bounding box.

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java:
```diff
@@ -401,6 +401,16 @@ public GeoBoundingBox getInternalBoundingBox() {
+            } catch (IOException e) {
+                // Most of the time the centroid is a good compromise, but perhaps we should know about this error anyway?
+                return centroid;
+            }
```
Not sure if we should ignore the error, maybe throw an UncheckedIOException instead?

iverase reviewed (commented):
Just have a small comment about how we handle an error. Otherwise let's get CI green and it can be approved.

craigtaverner committed:
Do not hide deserialization failures

craigtaverner committed:
Reduce depth of exception cause to pass tests

The changes to expose GeoShapeQueryable.toQuantizeLuceneGeometry to
external use also caused exception causes to be one level deeper.
This fix just brings the underlying message up one level to get
tests to pass.
While we could have also changed tests to assert on causes a level
deeper, this fix seems more backwards compatible, just in case anyone
else is relying on causes.

craigtaverner committed:
Polygon label position is centroid when within

iverase reviewed (approved):
LGTM

Thank you Craig!

stu-elastic reviewed (approved):
Scripting looks good.
